fix: Google Sheets datasource rehaul (#981)
appsmithorg/appsmith-docs: website/docs/reference/datasources/querying-google-sheets.md

---
sidebar_position: 7
toc_max_heading_level: 4
---

# Google Sheets

This page describes how to connect your application to your Google Sheets account and query your spreadsheets.

## Connect to Google Sheets

To add a Google Sheets datasource, click the (**+**) sign in the **Explorer** tab next to **Datasources**. On the next screen, select the **Google Sheets** button. Your datasource is created and you are taken to a screen to configure its settings.

## Configuration

Configure the Google Sheets Datasource as illustrated below:

### Scope

![](/img/google_sheets_scope.png)

The **Scope** setting defines what type of privileges your app has while querying your spreadsheets. Use this option to allow the minimum necessary privileges for your app's function to reduce the risk to your data.

> - **Read/Write | Selected Google Sheets**: Your app only has read and write access to the spreadsheets that it creates. Spreadsheets on the Google account created by other means aren't visible or mutable by your queries.

Author: > I don't think it has any relation to creation of spreadsheet.

Unfortunately, with the "Selected sheets" scope, I don't think there is any way to access a spreadsheet that your app has not created itself.

- **Read/Write | All Google Sheets**: Your app has read and write access to all spreadsheets on your Google account, regardless of how or when they were created.

- **Read Files | All Google Sheets**: Your app has read-only access to all spreadsheets on your Google account, regardless of how or when they were created.

> #### Save and authorize

Author: I don't quite remember the details of reasoning, but this came up in the weekly review and we kept it; I think it was to make it extra clear that the user **must** use this button and go through the flow or the datasource will not be saved.

Click on the **Save and Authorize** button once you have selected your **Scope**. You'll be directed to a Google Login screen, where you can log into the account whose spreadsheets you’d like to access.

> On a successful login, a screen appears for granting Appsmith permissions for managing your Google Sheets. Click "Allow" to allow Appsmith to manage your spreadsheets.

Reviewer: ```suggestion
Click **Allow** to grant the Appsmith app permission to view and edit files based on the scope.
```

## Create queries

<figure>

>   <img src="/img/google-sheets-query-page.png" style={{width: "100%", height: "auto"}} alt="Configuring a query from the query screen."/>

Reviewer: Maybe this image should show the dropdown with all the queries?

Author: that section was removed, so the image was moved somewhere that makes more sense for what is shown

  <figcaption align="center"><i>Configuring a query from the query screen.</i></figcaption>
</figure>

> You can create [queries](https://docs.appsmith.com/core-concepts/data-access-and-binding/querying-a-database/query-settings) to fetch or write data to your spreadsheets by clicking (**+**) next to **Queries/JS** in the **Explorer** tab and selecting your Google Sheets datasource. You'll be brought to a new screen to set up your query.

Reviewer: place the image below this para

### Operations

**Operation** sets the type of action you want to perform with your query. A "Spreadsheet" is a document, a "Sheet" is a page of a spreadsheet, and "Sheet Rows" are horizontal records in a sheet.

| **Operation**                        | **Description**                           | **Available entities:**       |
| ------------------------------------ | ----------------------------------------- | ----------------------------- |

> | [**Fetch Details**](#fetch-details)  | Fetches metadata about a spreadsheet. | [Spreadsheet](#fetch-details-spreadsheet)  |

Reviewer: Mention this is only accessible if you give Appsmith the read-write All permission

| [**Insert Many**](#insert-many)      | Inserts several new rows into a spreadsheet. | [Sheet Rows](#insert-many-sheet-rows) |
| [**Update Many**](#update-many)      | Updates multiple existing records in a spreadsheet.  | [Sheet Rows](#update-many-sheet-rows)  |

All the operation types have some of these common fields that identify where in your spreadsheets your query should access:

| **Configuration Field**     | **Description**                                                               |
| --------------------------- | ----------------------------------------------------------------------------- |
| **Entity**                  | Select which entity type you want to query. (Sheet Rows, Spreadsheet, Sheet). |
| **Spreadsheet**             | Select which spreadsheet you want to query from.                              |
| **Sheet Name**              | Select which sheet you want to query from the spreadsheet.                    |
| **Table Heading Row Index** | Provide the index of the row in the spreadsheet that contains the headings or labels for your table columns. The first row of the spreadsheet is row 1.                                                     |

> ### Fetch details

Reviewer: ```suggestion
### Fetch Details
```

Let's keep the text as it appears on the product

Use this operation to request metadata about one existing spreadsheet.

> ##### Spreadsheet {#fetch-details-spreadsheet}

Reviewer: As this is the only option, a heading here is not required

**Filter Format** provides two options for filtering your spreadsheet records:

The **Cell range** filter format selects a specific block of cells from the spreadsheet using Google Sheets' row number and column letter syntax (for example, `A1-B14`). Even when the column header row isn't part of your selection, your fetched data still includes the column labels for your selected cells.

Use the **Where Clause** filter format to filter based on your data. These fields enable you to return records conditionally, sort on multiple levels, and set up pagination.

In the **Filter by** fields, you can build expressions that return records only when a given column value meets some logical criteria. You can evaluate values using logical operators, "in," "not in," and "contains." The **Add Condition** button adds another simple single-line expression; **Add Group Condition** enables you to write a nested expression with multiple levels of And/Or statements.

In the **Sort By** field, you can choose a column to use for sorting your results. The **Add Parameter** button adds multiple levels of sorting.

> To limit the amount of records you receive at once, use **Pagination Limit**. **Pagination Offset** allows skipping a given number of records before returning results; these two fields together enable you to implement pagination for large datasets.

Reviewer: ```suggestion
To limit the amount of records you receive at once, use **Pagination Limit**. **Pagination Offset** allows skipping a given number of records before returning results; these two fields enable you to implement [serverside pagination](/reference/widgets/table#server-side-pagination) for large datasets.
```

---

**Example**: create a new record in a table `users` on `Sheet1` of `UsersSpreadsheet`, with columns for `name`, `gender`, and `email`.

Create your query called `InsertNewUser` based on your Google Sheets datasource. This query should use the **Insert One** operation for the **Sheet Rows** entity. Enter the appropriate values for **Spreadsheet**, **Sheet Name**, and **Table Heading Row Index**.

To gather data for the new record, you build a [Form](/reference/widgets/form) on the canvas called `NewUserForm` containing:

- An [input widget](/reference/widgets/input) called "NameInput" for the name,
- A [Select widget](/reference/widgets/select) called "GenderSelect" for the gender,
- An input widget called "EmailInput" for the email.

Once these form fields are filled out, you can add their values to your query like below:

```javascript
// in the Row Object field of your query
{{
  {
    "name": NewUserForm.data.NameInput,
    "gender": NewUserForm.data.GenderSelect,
    "email": NewUserForm.data.EmailInput
  }
}}

```

When your query is executed, the new record is inserted as the new highest index in your dataset (at the bottom of the spreadsheet).

> #### Insert many

Reviewer: ```suggestion
### Insert Many
```

Let it be same heading level
